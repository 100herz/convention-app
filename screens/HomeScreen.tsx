import React from 'react'
import { Button, StyleSheet, Text, View } from 'react-native'
import { StackNavigationProp } from '@react-navigation/stack'

import { RootStackParamList } from '@navigations/ArticleNavigator'

interface Props {
  navigation: StackNavigationProp<RootStackParamList, 'Categories'>
}

const HomeScreen: React.FC<Props> = ({ navigation }) => {
  return (
    <View style={styles.container}>
      <Text>Here comes the Articles component.</Text>
      <Text style={{ fontFamily: 'open-sans-bold', marginTop: 15, color: 'blue' }}>from HomeScreen.tsx</Text>
      <Button title="Go to Categories Screen" onPress={() => navigation.navigate('Categories')} />
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
  },
})

export default HomeScreen
