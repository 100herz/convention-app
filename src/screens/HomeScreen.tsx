import React, { useEffect, useState } from 'react'
import { SectionList, StyleSheet, View } from 'react-native'

import FeaturedCarousel from '@components/Articles/FeaturedCarousel'
import NewsCarousel from '@components/Articles/NewsCarousel'
import ArticleList from '@components/Articles/ArticleList'
import ArticlePreview from '@components/Articles/ArticlePreview'
import LoadingSpinner from '@components/UI/LoadingSpinner'
import Text from '@components/UI/Text'
import { useInfinityFetchHook } from '@hooks/use-infinity-fetch-hook'
import { Article } from '@models/article'
import { DefaultStyles, defaultStyles } from '@styles/theme'

const HomeScreen: React.FC = () => {
  const [initialLoading, setInitialLoading] = useState(true)
  const [newsArticles, setNewsArticles] = useState<Article[]>([])
  const [featuredSliderArticles, setFeaturedSliderArticles] = useState<Article[]>([])
  const [newsSliderArticles, setNewsSliderArticles] = useState<Article[]>([])
  const [pinnedArticle, setPinnedArticle] = useState<Article>()

  const { articles, fetchMore, isLoading } = useInfinityFetchHook()

  const isSponsored = (sponsoredProperty: string | null | undefined) =>
    sponsoredProperty !== undefined && sponsoredProperty !== null && sponsoredProperty.length > 0

  useEffect(() => {
    const filteredNewsArticle = articles.filter(article => article.categories.includes(1) && !article.acf?.pinned)

    if (filteredNewsArticle.length === 0) return

    if (isSponsored(filteredNewsArticle[0].acf?.sponsored_by)) filteredNewsArticle.shift()

    setNewsArticles(filteredNewsArticle)

    if (featuredSliderArticles.length > 0 || newsSliderArticles.length > 0 || pinnedArticle) return

    setFeaturedSliderArticles(articles.filter(article => article.acf?.featured_slider).slice(0, 5))
    setNewsSliderArticles(articles.filter(article => article.acf?.news_slider).slice(0, 10))
    setPinnedArticle(articles.find(article => article.acf?.pinned))
    setInitialLoading(false)
  }, [articles])

  const SectionHeader = () => (
    <View>
      <FeaturedCarousel articles={featuredSliderArticles} />
      <Text style={styles.title}>News</Text>
      <NewsCarousel articles={newsSliderArticles} />
      {pinnedArticle && <ArticlePreview article={pinnedArticle} />}
    </View>
  )

  return (
    <View style={styles.container}>
      {isLoading || initialLoading ? (
        <LoadingSpinner />
      ) : (
        <SectionList
          keyExtractor={item => item.id.toString()}
          stickySectionHeadersEnabled={false}
          sections={[{ data: newsArticles }]}
          renderItem={({ item }: { item: Article }) => <ArticlePreview article={item} />}
          renderSectionHeader={() => <SectionHeader />}
          initialNumToRender={10}
          onEndReachedThreshold={2}
          onEndReached={fetchMore}
        >
          <ArticleList data={newsArticles} />
        </SectionList>
      )}
    </View>
  )
}

const styles = StyleSheet.create<DefaultStyles>({
  ...defaultStyles,
  title: {
    ...defaultStyles.title,
    paddingHorizontal: 15,
    paddingTop: 25,
    paddingBottom: 15,
  },
})

export default HomeScreen
