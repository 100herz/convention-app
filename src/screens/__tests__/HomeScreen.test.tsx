import React from 'react'
import { fireEvent, render } from '@testing-library/react-native'

import HomeScreen from '../HomeScreen'

export const mockedNavigate = jest.fn()
jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native')
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
    useNavigationParam: jest.fn(jest.requireActual('@react-navigation/native').useNavigationParam),
  }
})

describe('<HomeScreen />', () => {
  it('renders correctly', () => {
    const { toJSON } = render(<HomeScreen />)
    expect(toJSON()).toMatchSnapshot()
  })

  it('should call the navigate function once', () => {
    const { getByTestId } = render(<HomeScreen />)
    fireEvent.press(getByTestId('button'))
    expect(mockedNavigate).toHaveBeenCalledTimes(1)
  })
})
