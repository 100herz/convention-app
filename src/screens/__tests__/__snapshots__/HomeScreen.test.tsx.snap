// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomeScreen /> renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <Text
    style={
      Object {
        "color": "blue",
        "fontFamily": "sans-bold",
        "marginVertical": 15,
      }
    }
  >
    from HomeScreen.tsx
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="button"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Go to Categories Screen
      </Text>
    </View>
  </View>
</View>
`;
